.docker/ci/Dockerfile

```diff
@@ -1,50 +1,90 @@
+    apt-get -qq dist-upgrade -y && \
     #
     # Install some base dependencies
     apt-get -qq install -y \
-        # Some source builds require a package.xml be downloaded via wget from an external location
         wget \
         # Required for rosdep command
         sudo \
-        # Preferred build tools
-        python-catkin-tools \
         clang clang-format-3.9 clang-tidy clang-tools \
-        ccache && \
-    #
-    # Download all dependencies of MoveIt!
+        ccache \
+        curl gnupg2 lsb-release \
+        # Some python dependencies for working with ROS2
+        python3-colcon-common-extensions \
+        python3-pip \
+        python-rosdep \
+        python3-wstool \
+        python3-rospkg-modules \
+        python3-rosdistro-modules \
+        && \
+    # Clear apt-cache to reduce image size
+    rm -rf /var/lib/apt/lists/*
+
+# Upgrade Pip and install some packages needed for testing
+# NOTE(mlautman): These pip installs are from the ros2 source install instructions. Seeing as not all of them
+#        are installed in the base image I added them here. I have not been able to verify that they are needed
+#        and if they aren't necessary they should be removed.
+RUN python3 -m pip install --upgrade pip && \
+    python3 -m pip install -U \
+        argcomplete \
+        flake8 \
+        flake8-blind-except \
+        flake8-builtins \
+        flake8-class-newline \
+        flake8-comprehensions \
+        flake8-deprecated \
+        flake8-docstrings \
+        flake8-import-order \
+        flake8-quotes \
+        # git+https://github.com/lark-parser/lark.git@0.7d \
+        pytest-repeat \
+        pytest-rerunfailures \
+        pytest \
+        pytest-cov \
+        pytest-runner \
+        setuptools
+
+# Install Fast-RTPS dependencies
+RUN apt-get -qq update && \
+    apt-get install -qq --no-install-recommends -y \
+        libasio-dev \
+        libtinyxml2-dev \
+        && \
+    # Clear apt-cache to reduce image size
+    rm -rf /var/lib/apt/lists/*
+
+# Download moveit source, so that we can get necessary dependencies
+RUN mkdir src && \
+    cd $ROS_WS/src && \
+    # TODO(mlautman): This should be changed to use vcs in the future
+    wstool init --shallow . https://raw.githubusercontent.com/ros-planning/moveit2/master/moveit.rosinstall
+
+# Download all MoveIt dependencies
+RUN \
+    apt-get -qq update && \
     rosdep update && \
     rosdep install -y --from-paths . --ignore-src --rosdistro ${ROS_DISTRO} --as-root=apt:false && \
-    #
-    # Remove the source code from this container
-    # TODO: in the future we may want to keep this here for further optimization of later containers
-    cd .. && \
-    rm -rf src/ && \
-    #
     # Clear apt-cache to reduce image size
     rm -rf /var/lib/apt/lists/*
 
+# Remove the source code from this container
+# TODO: in the future we may want to keep this here for further optimization of later containers
+RUN cd $ROS_WS && \
```

Comment:
The whole `ROS_WS` can be removed, right?

This command should be part of the same command downloading the MoveIt sources.
As @davetcoleman teached me a year? ago, commands creating and removing files/folders should be kept together in a single docker RUN command. For each RUN command a new image layer is added for all filesystem changes in this command. So, if we add and remove files in separate commands, we need to save and download the layer caches for both, additions and removals, while if both shell commands are handled in the same docker command, there are virtually no changes to save and download.

Reply:
In view of https://github.com/ros-planning/moveit2/pull/46#discussion_r268853434, I think keeping separate commands for apt-get is fine.
However, downloading MoveIt sources, installing dependencies via `rosdep`, and cleaning should all be handled in a single command then to minimize layer size.
Do you agree @ruffsl / @gavanderhoorn?

Reply:
> However, downloading MoveIt sources, installing dependencies via rosdep, and cleaning should all be handled in a single command then to minimize layer size.

Given that we are now cleaning up after ourselves after each RUN invocation and apt installation, there is little tangible difference in overall size, while at the same time we gain finer granularity in our build caching.

> The whole ROS_WS can be removed, right?

Acutely, I think we should omit this sanitization step entirely. It's rather valuable to have the exact source code on hand in the image that was used in building the binaries in the workspace. In navigation2, we use the existing source in the workspace baked into the image to seed the keys for later CI caching, as it worthwhile be able to determine if source code dependencies have changed from the latest pull. Cleaning the existing source I think is best left to the CI script that is to use the image, as we do via a pre_checkout step here:

https://github.com/ros-planning/navigation2/blob/bbbaf6d0cde05e89db871601677f134a823e13b3/.circleci/config.yml#L3-L11

Reply:
For this use case, we have a different docker container: `source`.
Here, the workspace is only needed to pull in dependencies!